contributors/design-proposals/local-ephemeral-storage.md

```diff
@@ -0,0 +1,580 @@
+          containers:
+          - name: nginx
+            image: gcr.io/google_containers/nginx-slim:0.8
+            ports:
+            - containerPort: 80
+              name: web
+            volumeMounts:
+            - name: www
+              mountPath: /usr/share/nginx/html
+            - name: log
+              mountPath: /var/log/nginx
+      volumeClaimTemplates:
+      - metadata:
+          name: www
+        spec:
+          accessModes: [ "ReadWriteOnce" ]
+          storageClassName: local-fast
+          resources:
+            requests:
+              storage: 100Gi
+      - metadata:
+          name: log
+        spec:
+          accessModes: [ "ReadWriteOnce" ]
+          storageClassName: local-slow
+          resources:
+            requests:
+              storage: 1Gi
+    ```
+
+4. The scheduler identifies nodes for each pod that can satisfy all the existing predicates.
+5. The nodes list is further filtered by looking at the PVC's StorageClass, and checking if there is available PV of the same StorageClass on a node.
+6. The scheduler chooses a node for the pod based on a ranking algorithm.
+7. Once the pod is assigned to a node, then the pod’s local PVCs get bound to specific local PVs on the node.
+
+    ```
+    $ kubectl get pvc
+    NAME            STATUS VOLUME     CAPACITY ACCESSMODES … NODE
+    www-local-pvc-1 Bound  local-pv-1 100Gi    RWO           node-1
+    www-local-pvc-2 Bound  local-pv-1 100Gi    RWO           node-2
+    www-local-pvc-3 Bound  local-pv-1 100Gi    RWO           node-3
+    log-local-pvc-1 Bound  local-pv-2 10Gi     RWO           node-1
+    log-local-pvc-2 Bound  local-pv-2 10Gi     RWO           node-2
+    log-local-pvc-3 Bound  local-pv-2 10Gi     RWO           node-3
+    ```
+    ```
+    $ kubectl get pv
+    NAME       CAPACITY … STATUS    CLAIM           NODE
+    local-pv-1 100Gi      Bound     www-local-pvc-1 node-1
+    local-pv-2 10Gi       Bound     log-local-pvc-1 node-1
+    local-pv-1 100Gi      Bound     www-local-pvc-2 node-2
+    local-pv-2 10Gi       Bound     log-local-pvc-2 node-2
+    local-pv-1 100Gi      Bound     www-local-pvc-3 node-3
+    local-pv-2 10Gi       Bound     log-local-pvc-3 node-3
+    ```
+
+8. If a pod dies and is replaced by a new one that reuses existing PVCs, the pod will be placed on the same node where the corresponding PVs exist. Stateful Pods are expected to have a high enough priority which will result in such pods preempting other low priority pods if necessary to run on a specific node.
+9. Forgiveness policies can be specified as tolerations in the pod spec for each failure scenario.  No toleration specified means that the failure is not tolerated.  In that case, the PVC will immediately be unbound, and the pod will be rescheduled to obtain a new PV.  If a toleration is set, by default, it will be tolerated forever.  `tolerationSeconds` can be specified to allow for a timeout period before the PVC gets unbound.
```

Comment:
Does this mean if a local PV fail, the pod will be "migrated" automatically to another node with another working local PV?

If so, I'm curious what is the reasoning for this? Why not just mark the pod as failure instead?

Reply:
I am not sure your question, but here the case is pod failed, and we have the options to continue to run on the previous node with the same PV, or schedule to a different node with different PV.

Reply:
> or schedule to a different node with different PV.

Does this happen automatically without human interaction?

Reply:
yes, statefulset will check if a pod dies, it restart a new pod automatically.

Reply:
What happens if a local PV fail? What would the pod become without any external action (e.g. stateful controller)?

Reply:
what do you mean by local PV fail?

Reply:
My fault. Let me clarify my thoughts and questions in details.

> Forgiveness policies can be specified as tolerations in the pod spec for each failure scenario.

I think this is not actually related to volume tolerations. At this moment, it can only tolerate node unready conditions. Note that node failure is different from volume failure. And this doc is about volume not about node.

Instead, I think this paragraph could be improved to talk about how to tolerate volume failure.

So on volume failure, my thought is that if the underlying PV has failed, it will have taint "storage.kubernetes.io/pvUnhealthy". Right? What happens happen? Will the related PVC change any status? Will the related pod change any status?

Reply:
Currently we don't have a health check for local disk. The plan is to add this function and if disk fails, PV is marked as "taint". And depending on the policy, the PVC will become unbound, then kubelet should check the status and evict the pod. Statefulset controller will create a new pod which will use a different PV. 

You could check the feature on local persistent volume https://github.com/kubernetes/community/pull/989

Reply:
That's great. Now I understand it.

> The plan is to add this function and if disk fails, PV is marked as "taint". And depending on the policy, the PVC will become unbound, then kubelet should check the status and evict the pod.

I think this information is important but missing in the docs. So it would be helpful to add this into the paragraph for folks reading this later.

Reply:
I think this is out of scope for this design doc.

Reply:
remove this section since it is not very related